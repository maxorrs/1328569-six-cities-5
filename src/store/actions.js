export const ActionType = {
  CHANGE_SELECTED_CITY: `CHANGE_SELECTED_CITY`,
  CHANGE_ACTIVE_CARD: `CHANGE_ACTIVE_CARD`,
  RESET_ACTIVE_CARD: `RESET_ACTIVE_CARD`
};

export const ActionCreator = {
  changeSelectedCity: (payload) => ({
    type: ActionType.CHANGE_SELECTED_CITY,
    payload
  }),
  changeActiveCard: (payload) => ({
    type: ActionType.CHANGE_ACTIVE_CARD,
    payload
  }),
  resetActiveCard: () => ({
    type: ActionType.RESET_ACTIVE_CARD,
    payload: `-1`
  })
};
