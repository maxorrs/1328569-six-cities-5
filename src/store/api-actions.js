import {DataActionCreator} from './reducers/data/data';
import {UserActionCreator} from './reducers/user/user';
import {AuthorizationStatus, getApiRoute, getAppRoute} from '../consts';
import browserHistory from '../browser-history';

export const fetchOffersList = () => (dispatch, _getState, api) => (
  api.get(getApiRoute.offers())
    .then(({data}) => {
      dispatch(DataActionCreator.toggleStatusOffers(true));
      dispatch(DataActionCreator.loadOffers(data));
      dispatch(DataActionCreator.toggleStatusOffers(false));
    })
);

export const fetchOffer = (id) => (dispatch, _getState, api) => (
  api.get(getApiRoute.offer(id))
    .then(({data}) => {
      dispatch(DataActionCreator.toggleStatusOffer(true));
      dispatch(DataActionCreator.loadOffer(data));
      dispatch(DataActionCreator.toggleStatusOffer(false));
    })
);

export const checkAuth = () => (dispatch, _getState, api) => {
  dispatch(UserActionCreator.loadAuthStatus(true));
  return api.get(getApiRoute.login())
    .then(({data}) => {
      dispatch(UserActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(UserActionCreator.setUserData(data));
      dispatch(UserActionCreator.loadAuthStatus(false));
    })
    .catch(() => {
      dispatch(UserActionCreator.loadAuthStatus(false));
    });
};

export const fetchOffersNearbyList = (id) => (dispatch, _getState, api) => (
  api.get(getApiRoute.nearby(id))
    .then(({data}) => {
      dispatch(DataActionCreator.toggleStatusOffersNearby(true));
      dispatch(DataActionCreator.loadOffersNearby(data));
      dispatch(DataActionCreator.toggleStatusOffersNearby(false));
    })
);

export const fetchFavoritesList = () => (dispatch, _getState, api) => (
  api.get(getApiRoute.favorites())
    .then(({data}) => {
      dispatch(DataActionCreator.toggleStatusFavorites(true));
      dispatch(DataActionCreator.loadFavorites(data));
      dispatch(DataActionCreator.toggleStatusFavorites(false));
    })
);

export const fetchReviews = (id) => (dispatch, _getState, api) => (
  api.get(getApiRoute.reviews(id))
    .then(({data}) => {
      dispatch(DataActionCreator.toggleStatusReviews(true));
      dispatch(DataActionCreator.loadReviews(data));
      dispatch(DataActionCreator.toggleStatusReviews(false));
    })
);

export const login = ({email, password}) => (dispatch, _getState, api) => {
  dispatch(UserActionCreator.checkedData(true));

  return api.post(getApiRoute.login(), {email, password})
    .then(({data}) => {
      dispatch(UserActionCreator.authDataHasError(false));
      dispatch(UserActionCreator.checkedData(false));
      dispatch(UserActionCreator.setUserData(data));
      dispatch(UserActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
    })
    .catch(() => {
      dispatch(UserActionCreator.authDataHasError(true));
      dispatch(UserActionCreator.checkedData(false));
    });
};

export const changeBookmarkStatus = (id, status) => (dispatch, _getState, api) => (
  api.post(getApiRoute.bookmark(id, status))
    .then(() => dispatch(fetchFavoritesList()))
    .catch(() => browserHistory.push(getAppRoute.login()))
);

export const sendReview = (id, {review: comment, rating}) => (dispatch, _getState, api) => {
  dispatch(DataActionCreator.toggleStatusSendReview(true));
  return api.post(getApiRoute.reviews(id), {comment, rating})
    .then(({data}) => {
      dispatch(DataActionCreator.sentReviewHasError(false));
      dispatch(DataActionCreator.loadReviews(data));
      dispatch(DataActionCreator.toggleStatusSendReview(false));
    })
    .catch(() => {
      dispatch(DataActionCreator.sentReviewHasError(true));
      dispatch(DataActionCreator.toggleStatusSendReview(false));
    });
};
