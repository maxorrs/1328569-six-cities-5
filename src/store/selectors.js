import {createSelector} from 'reselect';

import {NameSpace} from './reducers/root-reducer';
import {filterOffers, getFavoriteOffers} from '../utils/common';
import {getCityCoords, getOffersCoords} from '../utils/map';
import {adaptToClient} from '../utils/adapter';

export const getOffersSelector = (state) => state[NameSpace.DATA].offers;

export const getOffersAdaptSelector = createSelector(
    getOffersSelector,
    (offers) => offers.map(adaptToClient)
);

export const getActiveCardSelector = (state) => state[NameSpace.APP_STATE].activeCard;

export const getSelectedCitySelector = (state) => state[NameSpace.APP_STATE].selectedCity;

export const getSelectedSortTypeSelector = (state) => state[NameSpace.APP_STATE].selectedSortType;

export const getFilteredOffersSelector = createSelector(
    [getOffersAdaptSelector, getSelectedCitySelector, getSelectedSortTypeSelector],
    filterOffers
);

export const getOffersCoordsSelector = createSelector(
    getFilteredOffersSelector,
    getOffersCoords
);

export const getCityCoordsSelector = createSelector(
    [getOffersSelector, getSelectedCitySelector],
    getCityCoords
);

export const getFavoriteOffersSelector = createSelector(
    getOffersAdaptSelector,
    getFavoriteOffers
);

export const getSortMenuStatusSelector = (state) => state[NameSpace.APP_STATE].isSortMenuOpen;

export const getOffersNearbySelector = (state) => state[NameSpace.DATA].offersNearby;

export const getOffersNearbyAdaptSelector = createSelector(
    getOffersNearbySelector,
    (offers) => offers.map(adaptToClient)
);
