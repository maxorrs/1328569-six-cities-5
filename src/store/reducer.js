import {extend} from '../utils/common';
import {ActionType} from './actions';

import offers from '../mocks/offers';

const DEFAULT_CITY = `Amsterdam`;

const initialState = {
  selectedCity: DEFAULT_CITY,
  activeCard: `-1`,
  offers
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_SELECTED_CITY:
      return extend(state, {
        selectedCity: action.payload
      });

    case ActionType.CHANGE_ACTIVE_CARD:
      return extend(state, {
        activeCard: action.payload
      });

    case ActionType.RESET_ACTIVE_CARD:
      return extend(state, {
        activeCard: action.payload
      });
  }

  return state;
};

export {reducer};
