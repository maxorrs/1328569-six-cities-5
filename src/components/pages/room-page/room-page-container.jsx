import React, {memo, useEffect} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {compose} from 'redux';

import RoomPage from './room-page';

import {getOfferAdaptSelector, getOffersNearbyAdaptSelector, getStatusOfferSelector, getStatusOffersNearbySelector, getOffersCoordsSelector, getCityCoordsSelector} from '../../../store/reducers/data/selectors';
import {fetchOffersNearbyList, fetchOffer} from '../../../store/api-actions';
import {offerPropTypes} from '../../../utils/prop-types';

const RoomPageContainer = (props) => {
  const {idMatch, onNearbyOffersLoad, onOfferLoad} = props;

  useEffect(() => {
    onOfferLoad(idMatch);
    onNearbyOffersLoad(idMatch);
  }, [idMatch]);

  return <RoomPage {...props} />;
};

RoomPageContainer.propTypes = {
  offersNearby: PropTypes.arrayOf(offerPropTypes),
  onOfferLoad: PropTypes.func.isRequired,
  onNearbyOffersLoad: PropTypes.func.isRequired,
  idMatch: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired
};

const mapStateToProps = (state) => ({
  offer: getOfferAdaptSelector(state),
  isLoading: getStatusOfferSelector(state) || getStatusOffersNearbySelector(state),
  offersNearby: getOffersNearbyAdaptSelector(state),
  offersCoords: getOffersCoordsSelector(state),
  cityCoords: getCityCoordsSelector(state)
});

const mapDispathToProps = (dispatch) => ({
  onOfferLoad: (id) => {
    dispatch(fetchOffer(id));
  },
  onNearbyOffersLoad: (id) => {
    dispatch(fetchOffersNearbyList(id));
  }
});

export {RoomPageContainer};
export default compose(
    connect(mapStateToProps, mapDispathToProps),
    memo
)(RoomPageContainer);
