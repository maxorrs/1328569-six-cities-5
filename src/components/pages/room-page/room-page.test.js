import React from 'react';
import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import * as APIAction from '../../../store/api-actions';

import {RoomPageContainer} from './room-page-container';

const noop = () => {};

const NameSpace = {
  APP_STATE: `APP_STATE`,
  DATA: `DATA`,
  USER: `USER`
};

const getOffersAdaptMock = (count) => {
  const templateOffers = Array(count)
    .fill(``)
    .map((_, index) => {
      return {
        'city': {
          'name': `Amsterdam`,
          'location': {
            'latitude': 51.225402,
            'longitude': 6.776314,
            'zoom': 13
          }
        },
        'previewImage': `url`,
        'images': [`url1`, `url2`],
        'title': `Canal View Prinsengracht`,
        'isFavorite': Boolean(Math.random()),
        'isPremium': false,
        'rating': 4.7,
        'type': `Room`,
        'bedrooms': 1,
        'maxAdults': 3,
        'features': [
          {'field': `entire`, 'label': `Room`},
          {'field': `bedrooms`, 'label': `1 Bedrooms`},
          {'field': `adults`, 'label': `Max 3 adults`}
        ],
        'price': 298,
        'goods': [`Laptop friendly workspace`, `Breakfast`, `Washer`],
        'host': {
          'id': 25,
          'name': `Angelina`,
          'isPro': true,
          'avatarUrl': `img/avatar-angelina.jpg`
        },
        'description': `Description`,
        'location': {
          'latitude': 51.222402,
          'longitude': 6.786314,
          'zoom': 16
        },
        'id': index + 1
      };
    });

  if (count === 1) {
    const [oneOffer] = templateOffers;
    return oneOffer;
  }

  return templateOffers;
};

const userDataMock = {
  'avatar_url': `url`,
  'email': `sad@a.ru`,
  'id': 1,
  'is_pro': false,
  'name': `sad`
};

const reviewsMock = [
  {
    'comment': `Comment`,
    'date': `2020-10-30T19:03:49.647Z`,
    'id': 1,
    'rating': 3,
    'user': {
      'avatar_url': `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/4.jpg`,
      'id': 13,
      'is_pro': false,
      'name': `Zak`,
    }
  },
  {
    'comment': `Comment2`,
    'date': `2020-10-30T19:03:49.647Z`,
    'id': 2,
    'rating': 5,
    'user': {
      'avatar_url': `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/5.jpg`,
      'id': 12,
      'is_pro': true,
      'name': `Abc`,
    }
  }
];

const offersCoordsMock = [
  {id: 76, location: [52.37554, 4.9019759999999994], zoom: 16},
  {id: 77, location: [52.37554, 4.9019759999999994], zoom: 16},
  {id: 78, location: [52.37554, 4.9019759999999994], zoom: 16}
];

const cityCoordsMock = {
  city: `Amsterdam`,
  location: [52.385540000000006, 4.902976],
  zoom: 16
};

const mockState = {
  [NameSpace.DATA]: {
    favorites: getOffersAdaptMock(3),
    offers: getOffersAdaptMock(3),
    reviews: reviewsMock,
    offersNearby: getOffersAdaptMock(5),
    offer: getOffersAdaptMock(1),
    statusOffers: false,
    statusOffer: false,
    statusOffersNearby: false,
    statusFavorites: false,
    statusReviews: false,
    statusSendReview: false,
    sentReviewHasError: false
  },
  [NameSpace.APP_STATE]: {
    selectedCity: `Amsterdam`,
    activeCard: -1,
    isSortMenuOpen: false,
    selectedSortType: `Popular`
  },
  [NameSpace.USER]: {
    authorizationStatus: `AUTH`,
    userData: userDataMock,
    hasError: false,
    isDataChecked: false,
    isLoadAuthStatus: false
  }
};

jest.mock(`../../../store/api-actions`);
APIAction.fetchReviews = () => (dispatch) => dispatch(jest.fn());

it(`RoomPageContainer is rendered correctly`, () => {
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const store = mockStore(mockState);

  const props = {
    idMatch: `2`,
    isLoading: false,
    onOfferLoad: noop,
    onNearbyOffersLoad: noop,
    offer: getOffersAdaptMock(1),
    offersNearby: getOffersAdaptMock(3),
    offersCoords: offersCoordsMock,
    cityCoords: cityCoordsMock
  };

  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <RoomPageContainer {...props} />
          </BrowserRouter>
        </Provider>, {
          createNodeMock: () => document.createElement(`div`)
        }
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
