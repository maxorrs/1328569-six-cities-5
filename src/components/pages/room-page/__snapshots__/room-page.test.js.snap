// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomPageContainer is rendered correctly 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="/img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  sad@a.ru
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="url1"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="url2"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Canal View Prinsengracht
            </h1>
            <button
              className="property__bookmark-button property__bookmark-button--active button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              5
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              Room
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              1 Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 3 adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              298
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Laptop friendly workspace
              </li>
              <li
                className="property__inside-item"
              >
                Breakfast
              </li>
              <li
                className="property__inside-item"
              >
                Washer
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="/img/avatar-angelina.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                Description
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews ·
              <span
                className="reviews__amount"
              >
                2
              </span>
            </h2>
            <ul
              className="reviews__list"
            >
              <li
                className="reviews__item"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/4.jpg"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Zak
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "60%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    Comment
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2020-10-30"
                  >
                    October 2020
                  </time>
                </div>
              </li>
              <li
                className="reviews__item"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/5.jpg"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Abc
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "100%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    Comment2
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2020-10-30"
                  >
                    October 2020
                  </time>
                </div>
              </li>
            </ul>
            <form
              action="#"
              className="reviews__form form"
              method="post"
              onSubmit={[Function]}
            >
              <label
                className="reviews__label form__label"
                htmlFor="review"
              >
                Your review
              </label>
              <div
                className="reviews__rating-form form__rating"
              >
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  disabled=""
                  id="5-stars"
                  name="rating"
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value="5"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="5-stars"
                  title="perfect"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  disabled=""
                  id="4-stars"
                  name="rating"
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value="4"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="4-stars"
                  title="good"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  disabled=""
                  id="3-stars"
                  name="rating"
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value="3"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="3-stars"
                  title="not bad"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  disabled=""
                  id="2-stars"
                  name="rating"
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value="2"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="2-stars"
                  title="badly"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  disabled=""
                  id="1-stars"
                  name="rating"
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value="1"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="1-stars"
                  title="terribly"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
              </div>
              <textarea
                className="reviews__textarea form__textarea"
                disabled=""
                id="review"
                name="review"
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Tell how was your stay, what you like and what can be improved"
                value=""
              />
              <div
                className="reviews__button-wrapper"
              >
                <p
                  className="reviews__help"
                >
                  To submit review please make sure to set 
                  <span
                    className="reviews__star"
                  >
                    rating
                  </span>
                   and describe your stay with at least 
                  <b
                    className="reviews__text-amount"
                  >
                    50
                     characters
                  </b>
                  .
                </p>
                <button
                  className="reviews__submit form__submit button"
                  disabled={true}
                  type="submit"
                >
                  Submit
                </button>
              </div>
            </form>
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <div
          id="map"
        />
      </section>
    </section>
  </main>
  <div
    className="container"
  >
    <section
      className="near-places places"
    >
      <h2
        className="near-places__title"
      >
        Other places in the neighbourhood
      </h2>
      <div
        className="near-places__list places__list"
      >
        <article
          className="place-card near-places__place-card"
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="near-places__image-wrapper place-card__image-wrapper"
          >
            <a
              href="/offer/1"
              onClick={[Function]}
            >
              <img
                alt="Place image"
                className="place-card__image"
                height="200"
                src="url"
                width="260"
              />
            </a>
          </div>
          <div
            className=" place-card__info"
          >
            <div
              className="place-card__price-wrapper"
            >
              <div
                className="place-card__price"
              >
                <b
                  className="place-card__price-value"
                >
                  €
                  298
                </b>
                <span
                  className="place-card__price-text"
                >
                  / night
                </span>
              </div>
              <button
                className="place-card__bookmark-button button place-card__bookmark-button--active"
                onClick={[Function]}
                type="button"
              >
                <svg
                  className="place-card__bookmark-icon"
                  height="19"
                  width="18"
                >
                  <use
                    xlinkHref="#icon-bookmark"
                  />
                </svg>
                <span
                  className="visually-hidden"
                >
                  To bookmarks
                </span>
              </button>
            </div>
            <div
              className="place-card__rating rating"
            >
              <div
                className="place-card__stars rating__stars"
              >
                <span
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                />
                <span
                  className="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <h2
              className="place-card__name"
            >
              <a
                href="/offer/1"
                onClick={[Function]}
              >
                Canal View Prinsengracht
              </a>
            </h2>
            <p
              className="place-card__type"
            >
              Room
            </p>
          </div>
        </article>
        <article
          className="place-card near-places__place-card"
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="near-places__image-wrapper place-card__image-wrapper"
          >
            <a
              href="/offer/2"
              onClick={[Function]}
            >
              <img
                alt="Place image"
                className="place-card__image"
                height="200"
                src="url"
                width="260"
              />
            </a>
          </div>
          <div
            className=" place-card__info"
          >
            <div
              className="place-card__price-wrapper"
            >
              <div
                className="place-card__price"
              >
                <b
                  className="place-card__price-value"
                >
                  €
                  298
                </b>
                <span
                  className="place-card__price-text"
                >
                  / night
                </span>
              </div>
              <button
                className="place-card__bookmark-button button place-card__bookmark-button--active"
                onClick={[Function]}
                type="button"
              >
                <svg
                  className="place-card__bookmark-icon"
                  height="19"
                  width="18"
                >
                  <use
                    xlinkHref="#icon-bookmark"
                  />
                </svg>
                <span
                  className="visually-hidden"
                >
                  To bookmarks
                </span>
              </button>
            </div>
            <div
              className="place-card__rating rating"
            >
              <div
                className="place-card__stars rating__stars"
              >
                <span
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                />
                <span
                  className="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <h2
              className="place-card__name"
            >
              <a
                href="/offer/2"
                onClick={[Function]}
              >
                Canal View Prinsengracht
              </a>
            </h2>
            <p
              className="place-card__type"
            >
              Room
            </p>
          </div>
        </article>
        <article
          className="place-card near-places__place-card"
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="near-places__image-wrapper place-card__image-wrapper"
          >
            <a
              href="/offer/3"
              onClick={[Function]}
            >
              <img
                alt="Place image"
                className="place-card__image"
                height="200"
                src="url"
                width="260"
              />
            </a>
          </div>
          <div
            className=" place-card__info"
          >
            <div
              className="place-card__price-wrapper"
            >
              <div
                className="place-card__price"
              >
                <b
                  className="place-card__price-value"
                >
                  €
                  298
                </b>
                <span
                  className="place-card__price-text"
                >
                  / night
                </span>
              </div>
              <button
                className="place-card__bookmark-button button place-card__bookmark-button--active"
                onClick={[Function]}
                type="button"
              >
                <svg
                  className="place-card__bookmark-icon"
                  height="19"
                  width="18"
                >
                  <use
                    xlinkHref="#icon-bookmark"
                  />
                </svg>
                <span
                  className="visually-hidden"
                >
                  To bookmarks
                </span>
              </button>
            </div>
            <div
              className="place-card__rating rating"
            >
              <div
                className="place-card__stars rating__stars"
              >
                <span
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                />
                <span
                  className="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <h2
              className="place-card__name"
            >
              <a
                href="/offer/3"
                onClick={[Function]}
              >
                Canal View Prinsengracht
              </a>
            </h2>
            <p
              className="place-card__type"
            >
              Room
            </p>
          </div>
        </article>
      </div>
    </section>
  </div>
</div>
`;
